### ISSUE TITLE
Epinio fails to find deployed applications in Rancher after its previous uninstall and reinstall

### ISSUE THREAD

**Issue by mmartin24** [epinio/epinio#2623]
# Issue

It was noted after https://github.com/epinio/epinio/issues/2349#issuecomment-1745065888, that a deployed app on namespace `workspace` existing in Rancher will survive an Epinio uninstall and re-install but Epinio fails to find it.

# Steps to reproduce:
- Deploy Rancher (from Epinio repo just run `make acceptance-cluster-delete && make acceptance-cluster-setup && export KUBECONFIG=$PWD/tmp/acceptance-kubeconfig && make install-cert-manager && make install-rancher `
#### Pls do the following steps via UI:
- Install the default Epinio chart (aka Marketplace Integration Chart) 
- Follow installation steps to deploy Epinio
- Create namespace `workspace`
- Push an app on that workspace
- Create a service and bind it to that app
- Uninstall Epinio and CRD using ui
- Reinstall the same Epinio chart

#### This ones to be done cli:
- Run `kubectl get all -n workspace`
- Run `epinio app list`

# Observed behavior

The `kubectl get all -n workspace` command will display the existing application as in:
```
~> k get all -n workspace
NAME                                                                  READY   STATUS    RESTARTS   AGE
pod/rsample-8151325dcdbae9e0ff95f9f9658432dbedfdb209-6b446c7f96svbm   1/1     Running   0          13m
pod/xc11fadea2478b7bb684678d6d8a8-mysql-0                             1/1     Running   0          13m

NAME                                                       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/rsample-8151325dcdbae9e0ff95f9f9658432dbedfdb209   ClusterIP   10.43.249.181   <none>        8080/TCP   16m
service/xc11fadea2478b7bb684678d6d8a8-mysql-headless       ClusterIP   None            <none>        3306/TCP   13m
service/xc11fadea2478b7bb684678d6d8a8-mysql                ClusterIP   10.43.223.219   <none>        3306/TCP   13m

NAME                                                               READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/rsample-8151325dcdbae9e0ff95f9f9658432dbedfdb209   1/1     1            1           16m

NAME                                                                          DESIRED   CURRENT   READY   AGE
replicaset.apps/rsample-8151325dcdbae9e0ff95f9f9658432dbedfdb209-6b446c7f97   1         1         1       13m
replicaset.apps/rsample-8151325dcdbae9e0ff95f9f9658432dbedfdb209-66c7d45bf    0         0         0       16m

NAME                                                   READY   AGE
statefulset.apps/xc11fadea2478b7bb684678d6d8a8-mysql   1/1     13m
```

The  `epinio app list` command though will fail to show the app in the targeted namespace:
```
epinio app list

:ship:  Listing applications
Namespace: workspace

:warning:  Epinio server version (v1.9.0) doesn't match the client version (v1.10.0)

:warning:  Update the client manually or run `epinio client-sync`

:heavy_check_mark:  Epinio Applications:
| NAME | CREATED | STATUS | ROUTES | CONFIGURATIONS | STATUS DETAILS |
|------|---------|--------|--------|----------------|----------------|
```

Services do seem however to be correctly found in the named namespace.

This behavior is also observed with Rancher:

![Image](https://github.com/epinio/epinio/assets/37271841/5e76d95f-9d97-4807-8f59-55eb5dd4e555)

![Image](https://github.com/epinio/epinio/assets/37271841/1ab165f2-ffe0-4657-871d-43365b5983cb)

![Image](https://github.com/epinio/epinio/assets/37271841/2d92b919-eec1-40ec-9c40-cc3e0d4f1157)

**Comment by thehejik** [comment]
I wonder from where is the application list in Epinio actually loaded. I guess for services and namespaces it is driven by k8s resource labels/annotations but this is probably not the case for Epinio Apps. 

It is true that by uninstalling Epinio we also get rid of the app data store (minio) but it shouldn't be required for app listing and/or their updating.

**Comment by andreas-kupries** [comment]
> Create namespace `workspace`

This specific namespace is actually created by the helm chart.
See `helm-charts/chart/epinio/templates/default-app-namespace.yaml`.

Uninstalling E should have taken down that workspace, as a chart-controlled resource.

It did when I uninstalled E from the local k3s cluster. Via `helm uninstall -n epinio epinio`.

However I had not created anything in the namespace yet. No app, no service.
Helm might not take it down if there are non-helm resources in the space ?
Checked, and false.
Doing a `helm uninstall -n epinio epinio` deletes namespace `workspace` and everything in it.

Even kube is not showing the namespace anymore, nor the app.
So this is not a reproduction, as the original issue had kube seeing things, and E not.

Seems that I have to add `Rancher` into the mix.

**Comment by andreas-kupries** [comment]
... Rancher setup looks to take longer ... (`make install-rancher`).
Lots of `helm-operation-...` pods around, some completed, some in error states.
To hopefully resolve themselves in time.
Although not very optimistic, the troubled pods are already 15min old.

Given that `workspace` is a special namespace (*) could you retry things with a different namespace of any other name ?

Should the local rancher not work I will go back to plain k3s, and then add a service to the namespace as well.
I am wondering if the service blocked the deletion somehow, and left app, etc. in some partial state.

Will try that for `workspace`, and some other namespace.

**Comment by andreas-kupries** [comment]
Another 10 min, rancher shows no change, no resolution to errors:
````
cattle-system                    rancher-77865fdc9d-hrms6                                         1/1    Running    2         (29m  ago)  33m
cattle-system                    rancher-77865fdc9d-cmgpt                                         1/1    Running    1         (29m  ago)  33m
cattle-system                    rancher-77865fdc9d-6lnh5                                         1/1    Running    1         (29m  ago)  33m
cattle-fleet-system              fleet-controller-6585bfdbfd-vqczd                                1/1    Running    0         28m         
cattle-system                    helm-operation-xl6pl                                             0/2    Completed  0         27m         
cattle-system                    helm-operation-qwhdk                                             1/2    Error      0         27m         
cattle-system                    helm-operation-qvtzq                                             1/2    Error      0         27m         
cattle-system                    helm-operation-ld2rh                                             1/2    Error      0         27m         
cattle-fleet-system              gitjob-789866b9b6-87r7t                                          1/1    Running    0         28m         
cattle-system                    helm-operation-4bdp7                                             1/2    Error      0         27m         
cattle-system                    helm-operation-hgpms                                             1/2    Error      0         27m         
cattle-system                    helm-operation-zx658                                             0/2    Completed  0         28m         
cattle-system                    rancher-webhook-76875cc6f4-9sf82                                 1/1    Running    0         27m         
cattle-system                    helm-operation-fjmzf                                             0/2    Completed  0         27m         
cattle-provisioning-capi-system  capi-controller-manager-7df4c77bcb-wtcvv                         1/1    Running    0         26m         
cattle-system                    helm-operation-nqxtf                                             0/2    Completed  0         26m         
```

Killing it now, trialing more without it.

**Comment by andreas-kupries** [comment]
Retried with app and service, bound together in the namespace.
Doing a `helm uninstall -n epinio epinio` did remove everything again.
Including everything in the namespace `workspace`.
__However__ this time it took a minute for the application to be fully terminated and removed.

This gives us ample time to look at resources with `kubectl` and see the app listed still.
And by the time `E` is reinstalled it likely was completely gone, leading to `E` not seeing it.

In the original report `kubectl get all -n workspace` was run, then `E` reinstalled, and then `epinio app list`.
When the last command failed to list the app, was `kubectl get all -n workspace` run again to verify that
it was still present in the cluster ?

**Comment by andreas-kupries** [comment]
Trialed with a different namespace, i.e. not `workspace`.
Everything looks good.
As expected the non-special namespace is not deleted as E is uninstalled.

And after reinstalling E the app and service in that namespace are properly visible.
After targeting that namespace, of course. Or using `--all`.

I suspect that in the original report, with the special `workspace`,
both app and service were deleted properly,
however taking enough time to still see them with an immediate kubectl command.
I suspect that checking for them with `kubectl` after reinstalling E would see them missing then too.
Not just missing in E.

@mmartin24 Please recheck.

**Comment by mmartin24** [comment]
It was clarified in the huddle meeting.
There was confusion about what the "default" chart meant. I meant it as Marketplace Integration Chart (I have already corrected it in the description), whereas it seems to have been understood to be the Epinio upstream chart which needs the git repo pointing to E. The first one does not ship the namespace `workspace` whereas the second does.

Also, the steps to reproduce this issue at least in the deployment part were done using the UI, which perhaps was understood on my side, but not explicitly said in the description. Changed as well.

**Comment by mmartin24** [comment]
Reproduced with Epinio 1.10.0 chart (present in Rancher 2.8.0-rc3) and I keep observing this issue. Attaching screenshot where it can be seen that `sample-app` is empty on cli while it appears on Rancher and Epinio UI

![image](https://github.com/epinio/epinio/assets/37271841/330d6577-765a-4f5c-95f4-88b7506a0666)

To install the Epinio `1.10.0` Rancher market integration chart I had to deploy Rancher version `v2.8.0-rc3`
For that, I modified the install-rancher.sh script adding the `--version v2.8.0-rc3` here:
```
helm upgrade --install rancher rancher-latest/rancher \
  --namespace cattle-system --create-namespace \
  --set global.cattle.psp.enabled=false \
  --set "hostname=$EPINIO_SYSTEM_DOMAIN" \
  --version v2.8.0-rc3 \
  --set bootstrapPassword=password \
  --wait
```

@andreas-kupries, In case it helps I recorded [this](https://mysuse-my.sharepoint.com/:v:/g/personal/manuel_martin_suse_com/EUPS3ImEyxNIjRBOjUdnKggBWLOnUFMU1ncE4WGsoZmqHg?nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJPbmVEcml2ZUZvckJ1c2luZXNzIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXciLCJyZWZlcnJhbFZpZXciOiJNeUZpbGVzTGlua0RpcmVjdCJ9fQ&e=cSrPEv) video with the chart installation, epinio push and later uninstall. I am happy to pair to show this in case it is needed.

**Comment by andreas-kupries** [comment]
Trying on my own first.

**Comment by andreas-kupries** [comment]
In the epinio checkout ...
Commit is main head (17e99da0d7aa9b149545742b520ead2742633e2a)
Running

> ( make acceptance-cluster-delete && make acceptance-cluster-setup && export KUBECONFIG=$PWD/tmp/acceptance-kubeconfig && make install-cert-manager && make install-rancher ) |& tee SETUP

Looks ok. Despite the `deadline exceeded` at the end all of the pods look to be running.
Able to log into Rancher dashboard ...

Install E 1.10.0, domain `172.18.0.4.omg.howdoi.website`, everything else defaults.
Able to log into the E dash board.

- Created namespace `lair` :heavy_check_mark: 
- Created service `foxhole` / mysql-dev :heavy_check_mark: 
- Created app `fox`, from `assets/sample-app` :heavy_check_mark:
- Bind `fox` and `foxhole` :heavy_check_mark: 
- Inspected from E cli, and K cli, saved results.
- Deleted E chart :heavy_check_mark: 
- Inspected from K cli, saved results.
- Reinstalled E chart  :heavy_check_mark: 
- E UI dashboard shows everything :heavy_check_mark: 
- Inspected from E cli, and K cli, saved results.
- Comparing results
- No differences :warning: 
- In other words, the application, and service were properly visible after E reinstall.

Summary: still failed to repro.

We should huddle tomorrow where you demonstrate your sequence and I can see where you and I do different stuff.

**Comment by andreas-kupries** [comment]
I can now fully explain the behaviour. It hinges on the epinio CRDs.
E comes with 3 CRDs, for app charts, service catalog entries, and apps.
The latter is the basis for the main resource containing the __description of a created app__.
Note that the __active parts__ of an app are __regular kube resources__, i.e. `Deployment`, etc.

When E is uninstalled from rancher, it is possible to uninstall `epinio` and `epinio-crd` separately.
The issue happens if/when `epinio-crd` is uninstalled.

- When a CRD is removed kube removes all associated CRs.
- In our case the app charts, service catalog entries, and the app descriptions.
- The active parts of apps are not removed, and the app keeps running.
- This means that a running app keeps being visible as such in the __Rancher__ dashboard.
- After reinstalling Epinio the default appchart CRs and service catalog CRs are restored.
- __Not restored__ are any user-defined appcharts, nor any user-defined service catalog entries.
- And, of course, the app descriptions are __not restored__ either.
- This is why __neither Epinio CLI, nor Epinio dashboard__ will see the app (description).
- It will still happily run, and be visible in the Rancher dashboard, consisting of regular resources.
- Any service instances are still visible because their main resource is a helm chart, whose CR/CRD is not under E's control, and not removed when E is uninstalled.

This is all pretty much as operating as designed.

If we wish for the app descriptions to survive we have to give up the `app CRD` with its CRs, and save our apps in config maps,  or secrets, or other suitable standard kube resource.

**Comment by mmartin24** [comment]
Thanks @andreas-kupries ,

ok, so it seems expected behavior given the actual configuration with the Rancher chart (not the upstream one).
I guess there is no need at the moment to change this, right? 

I was checking this with Tomas and IIUC what you suggest if we save for example an app `jenkins-epinio` from `apps.application.epinio.io` we could reuse it later for this purpose, correct? Like this : 
```
kubectl get apps.application.epinio.io jenkins-epinio -n workspace -o yaml
apiVersion: application.epinio.io/v1
kind: App
metadata:
  annotations:
    epinio.io/created-by: admin
  creationTimestamp: "2023-11-16T12:01:44Z"
  generation: 3
  name: jenkins-epinio
  namespace: workspace
  resourceVersion: "1847"
  uid: 60969d1b-9d8c-45f5-9faf-ad0fb993e695
spec:
  chartname: standard
  imageurl: jenkins/jenkins
  origin:
    container: jenkins/jenkins
  routes:
  - jenkins-epinio.192.168.16.3.omg.howdoi.website
```
Being this the case, would we want to document on the chart about this in case anyone finds this case using cli and Rancher?

**Comment by andreas-kupries** [comment]
>  I guess there is no need at the moment to change this, right?

That is my opinion, yes.

> expected behavior given the actual configuration with the Rancher chart (not the upstream one).

While you see it clearly with the rancher  chart, this can happen with the upstream chart as well.
The moment the CRDs are removed things break. And you can do that for the upstream chart too.

> kubectl get apps.application.epinio.io jenkins-epinio -n workspace -o yaml

Saving the descriptions. That makes sense. After the reinstall they can be restored.
We should document this in general, not just for Rancher chart.
Unclear where we could document it for the Rancher chart.

**Comment by mmartin24** [comment]
Closing issue. 
Opened this other one to address the documentation part: https://github.com/epinio/docs/issues/375
